sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -135,7 +135,9 @@ KernelProgramCache::BuildResult<RetT> *
 getOrBuild(KernelProgramCache &KPCache, GetCachedBuildFT &&GetCachedBuild,
            BuildFT &&Build) {
   using BuildState = KernelProgramCache::BuildState;
+  int n_retry = 0;
 
+out:
```

Comment:
I am not a huge fan of the `goto` approach as I think it makes the control-flow hard to follow for the reader. Could we maybe instead move the body of this function to a new function and make this function do something like

```c++
for (size_t AttemptCounter = 0; AttemptCounter < MaxBuildRetry; ++AttemptCounter) {
  KernelProgramCache::BuildResult<RetT> *Result = MyNewFunction(..., /*EvictOnOutOfResources=*/true);
  if (Result ->State.load() != BuildState::BS_WillEvict)
    return Result;
}
return MyNewFunction(..., /*EvictOnOutOfResources=*/false);
```
For now `MaxBuildRetry` will just be 1 as we evict the entire cache, but in the future we can change that behavior.

The new `EvictOnOutOfResources` argument would be used instead of `n_retry == 1` further down.